File: java/components/StrataPersistence/src/strata1/persistence/repository/IRepositoryContext.java
// ##########################################################################
// # File Name:	.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.persistence.repository;


import strata1.common.synchronizer.ISynchronizer;

/**
 * The base class of all repository contexts. Participates in the 
 * <a href="{@docRoot}/designpatterns/DomainObjectManager.pdf">RepositoryImp 
 * (aka Domain Object Manager)</a> design pattern.
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
interface IRepositoryContext
{
	/************************************************************************
	 * Returns the synchronizer associated with this 
	 * {@code IRepositoryContext} that synchronizes access to all
	 * of the {@code IRepositoryProvider}s associated with this 
	 * {@code IRepositoryContext}.
	 *
	 * @return synchronization mechanism
	 */
	public ISynchronizer 
	getSynchronizer();
	   
    /************************************************************************
     * Returns the current unit of work associated with this
     * {@code IRepositoryContext} that manages the transactions to all
     * of the {@code IRepositoryProvider}s associated with this
     * {@code IRepositoryContext}.
     * 
     * @see     IUnitOfWork
     * @return  current unit of work
     */
    public IUnitOfWork
    getUnitOfWork();    
}


// ##########################################################################

File: legacy/csharp/components/StrataPersistence/Repository/IRepositoryContext.cs
//  ##########################################################################
//  # File Name: IRepositoryContext.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

namespace Strata.Persistence.Repository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// Root interface of all repository contexts. Provides context-level
    /// locking and access to the current unit of work.
    /// </summary>
    /// 
    /// <author>JFL</author>
    /// <conventions>$conventionspath$</conventions>
    ///  
    public
    interface IRepositoryContext
    {
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Acquires a read lock on the context. 
        /// </summary>
        /// 
        void
        LockForReading();

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        ///  Releases a read lock on the context.
        /// </summary>
        /// 
        void
        UnlockFromReading();

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Acquires a write lock on the context.
        /// </summary>
        /// 
        void
        LockForWriting();

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Releases a write lock on the context. 
        /// </summary>
        /// 
        void
        UnlockFromWriting();

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Returns the current unit of work for this context. 
        /// </summary>
        /// 
        /// <returns>current unit of work</returns>
        /// 
        IUnitOfWork
        GetUnitOfWork();
   }
}

//  ##########################################################################
